### ISSUE TITLE
bug in fire model related to grass leaf fire consumption

### ISSUE THREAD

**Issue by ckoven** [NGEET/fates#412]
All,
In setting up regional FATES runs over California, I have been running into a strange bug where the model would crash when I turned on the spitfire.  The crash showed up as a NaN in the NPP term, as below:

```
42: Plant Growth Integrator could not find
42: a solution in less than          300  tries
42: Aborting
42: carbon_balance                     NaN
42: deltaC                     NaN
42: totalC                     NaN
42: leaf: T                     NaN  7.091519147632314E-002                     NaN
42: froot: T                     NaN  7.091519147632314E-002
42:                     NaN
42: sap: T                     NaN  3.530199739476804E-004  3.530470578426043E-004
42: store: T                     NaN  8.509823315308761E-002
42:                     NaN
42: dead:                     NaN  2.345192039175753E-002  2.345436637946122E-002
42: ----------------------------------------
42:  Dumping Cohort Information             
42: ----------------------------------------
42: co%pft                    =            3
42: co%n                      =    2.71210026657635     
42: co%dbh                    =    1.35295799649237     
42: co%hite                   =    2.84460496513949     
42: co%bdead                  =   2.345436637946122E-002
42: co%bstore                 =                      NaN
42: co%laimemory              =   0.000000000000000E+000
42: co%bsw                    =   3.530470578426043E-004
42: co%bl                     =                      NaN
42: co%br                     =                      NaN
42: co%lai                    =   0.000000000000000E+000
42: co%sai                    =   8.692855374800612E-003
42: co%g_sb_laweight          =   0.000000000000000E+000
42: co%leaf_cost              =   0.000000000000000E+000
42: co%canopy_layer           =            1
42: co%canopy_layer_yesterday =    1.00000000000000     
42: co%nv                     =            1
42: co%status_coh             =            2
42: co%canopy_trim            =   0.832085051966768     
42: co%excl_weight            =    4.39088230981449     
42: co%prom_weight            =   1.074470174389356E-010
42: co%size_class             =            1
42: co%size_by_pft_class      =           27
42: co%gpp_acc_hold           =                      NaN
42: co%gpp_acc                =                      NaN
42: co%gpp_tstep              =                      NaN
42: co%npp_acc_hold           =                      NaN
42: co%npp_tstep              =                      NaN
42: co%npp_acc                =                      NaN
42: co%resp_tstep             =                      NaN
42: co%resp_acc               =                      NaN
42: co%resp_acc_hold          =                      NaN
42: co%npp_leaf               =                      NaN
42: co%npp_fnrt              =                      NaN
42: co%npp_sapw               =   0.000000000000000E+000
42: co%npp_dead              =   0.000000000000000E+000
42: co%npp_seed              =   0.000000000000000E+000
42: co%npp_stor              =                      NaN
42: co%ode_opt_step          =                      NaN
42: co%rdark                  =   0.000000000000000E+000
42: co%resp_m                 =   4.954221740545094E-006
42: co%resp_g                 =                      NaN
42: co%livestem_mr            =   0.000000000000000E+000
42: co%livecroot_mr           =   0.000000000000000E+000
42: co%froot_mr               =   2.752345411413941E-009
42: co%md                     =   0.000000000000000E+000
42: co%leaf_md                =   0.000000000000000E+000
42: co%root_md                =   6.603636589486625E-002
42: co%bstore_md              =   0.000000000000000E+000
42: co%bdead_md               =   0.000000000000000E+000
42: co%bsw_md                 =   0.000000000000000E+000
42: co%dmort                  =   0.000000000000000E+000
42: co%seed_prod              =   0.000000000000000E+000
42: co%treelai                =   0.000000000000000E+000
42: co%treesai                =   0.218638733473752     
42: co%leaf_litter            =   0.000000000000000E+000
42: co%c_area                 =    2.63899866073878     
42: co%cmort                  =   0.000000000000000E+000
42: co%bmort                  =   0.000000000000000E+000
42: co%fmort                  =   0.000000000000000E+000
42: co%hmort                  =   0.000000000000000E+000
42: co%frmort                 =   0.000000000000000E+000
42: co%isnew                  =  F
42: co%dndt                   =   0.000000000000000E+000
42: co%dhdt                   =   0.000000000000000E+000
42: co%ddbhdt                 =   0.000000000000000E+000
42: co%dbdeaddt               =   0.000000000000000E+000
42: co%dbstoredt              =   0.000000000000000E+000
42: co%cfa                    =   0.000000000000000E+000
42: co%fire_mort              =   0.000000000000000E+000
42: co%crownfire_mort         =   0.000000000000000E+000
42: co%cambial_mort           =   0.000000000000000E+000
42: co%size_class             =            1
42: co%size_by_pft_class      =           27
42: ----------------------------------------
42: ENDRUN:
42: ERROR in EDPhysiologyMod.F90 at line 1371                                      
42:                                                                                
42:                                                                                
42:                                                                                
42:                                                                                
42:                                                                                
42:                                       
42: ERROR: Unknown error submitted to shr_abort_abort.
```

and if debug was set to true in the xml files, it would catch it as such:

```
85:forrtl: error (72): floating overflow
85:Image              PC                Routine            Line        Source             
85:cesm.exe           0000000003FCB6C1  Unknown               Unknown  Unknown
85:cesm.exe           0000000003FC97FB  Unknown               Unknown  Unknown
85:cesm.exe           0000000003F7B2F4  Unknown               Unknown  Unknown
85:cesm.exe           0000000003F7B106  Unknown               Unknown  Unknown
85:cesm.exe           0000000003EFABE9  Unknown               Unknown  Unknown
85:cesm.exe           0000000003F071DC  Unknown               Unknown  Unknown
85:libpthread-2.19.s  00002AAAAFAC1870  Unknown               Unknown  Unknown
85:cesm.exe           00000000017ABD4A  fatesplantresppho        1316  FatesPlantRespPhotosynthMod.F90
85:cesm.exe           00000000017A8E0E  fatesplantresppho         911  FatesPlantRespPhotosynthMod.F90
85:cesm.exe           000000000179C64A  fatesplantresppho         440  FatesPlantRespPhotosynthMod.F90
85:cesm.exe           0000000000977587  clmfatesinterface        1643  clmfates_interfaceMod.F90
85:cesm.exe           00000000011AE969  canopyfluxesmod_m         842  CanopyFluxesMod.F90
85:cesm.exe           0000000000887DA8  clm_driver_mp_clm         543  clm_driver.F90
85:cesm.exe           000000000084B5B9  lnd_comp_mct_mp_l         451  lnd_comp_mct.F90
85:cesm.exe           000000000046BD2D  component_mod_mp_         688  component_mod.F90
85:cesm.exe           000000000043C474  cime_comp_mod_mp_        2652  cime_comp_mod.F90
85:cesm.exe           00000000004543B7  MAIN__                     68  cime_driver.F90
85:cesm.exe           0000000000415A5E  Unknown               Unknown  Unknown
85:libc-2.19.so       00002AAAB190AB25  __libc_start_main     Unknown  Unknown
85:cesm.exe           0000000000415969  Unknown               Unknown  Unknown
-1:MPT ERROR: MPI_COMM_WORLD rank 85 has terminated without calling MPI_Finalize()
-1:	aborting job
MPT: Received signal 6
```

In discussion with @rgknox and @jkshuman, I tried a bunch of things to isolate the cause of the problem.  The proximal cause was an error in the quadratic calculation of the photosynthesis code, but I didn't want to just bound that and move along.  So I got it to pass out cohort diagnostics at the moment of the crash, and saw the following info:

```
42: ----------------------------------------
42:  Dumping Cohort Information             
42: ----------------------------------------
42: co%pft                    =            3
42: co%n                      =    58.8041129643066     
42: co%dbh                    =    1.47279463919445     
42: co%hite                   =    2.99973646208669     
42: co%bdead                  =   2.854267342198438E-002
42: co%bstore                 =   1.162382650830486E-002
42: co%laimemory              =   0.000000000000000E+000
42: co%bsw                    =   4.196605842665105E-004
42: co%bl                     =   6.942857596139393E-308
```

which suggested that something strange was happening to get such teeny leaf biomass.  Figuring that there must be some interaction with fire and leaf biomass, given the above clues, I started playing around with the code that calculates leaf consumption by fire.  @rgknox noticed that there were some untyped reals in the calculation of area burned, so I tried fixing those in 0333260, but that didn't fix the problem.  What did fix the problem (or seems to anyway) is changing the calculation so that there is no sapwood term in the leaf fire consumption calculation, in 7ad9216.    I'm not sure I understand why sapwood is part of this calculation, so don't know if removing it creates some theoretical problem.

So I wanted to ask if that calculation of the leaf fire consumption is correct in its current form, if the changes to 7ad9216 are better, or if some alternate solution is preferred.  I know that both @rosiealice and @jkshuman are currently on holiday but want to get their take on this when they return.

**Comment by jkshuman** [comment]
Had a conversation with @rgknox about this issue to discuss stemwood burning. I summarize our discussion and further thoughts here. 

Couple of thoughts:  Currently we damage the crown and remove leaves according to the proportion of the crown that is scorched, which then lowers photosynthetic capacity. Consumption of branches would burn sapwood and remove stemwood. This fire consumption differs from branch fire damage through cavitation and altered flow (see issue #413 ). I discuss here branch consumption by fire, but, maybe, this is best handled in conjunction with issue #413? I am inclined to leave this section as a flux of leaves at the moment and do more research on branch consumption for living trees. 

https://github.com/NGEET/fates/blob/ac07ad4b9c2c07ae6ef0f5cf085b82848b4f33e9/biogeochem/EDPatchDynamicsMod.F90#L989
The fire litter flux section in EdPatchDynamicsMod.F90 is for burning parts of trees that were not consumed in the fire and calculating the flux to the atmosphere. For trees, this involves removing leaves according to 'cfa' (agree with @rosiealice that we should rename this to fraction_crown_burned). Removing sapwood and stemwood would occur if twigs or branches are consumed by the fire. The actual trunk of the tree is not removed as this tree cohort has been evaluated in the fire routine and survived the combined probability of cambial and crown scorch damage.

For removal of branches we have the cfa term to determine how far into the tree canopy the fire travels, but then we must also account for fire residence time similar to cambial damage to the stem. Different size branches have different burn times, and fire consumption depends on fuel moisture. (In reality, flammability of live and dead fuels varies, so will need to address that eventually (see Jolly and Johnson 2018 Fire)). If branches are consumed, @rgknox suggested use of the trimming function to match the new crown depth per the change imposed by the cfa term. This line is in the fire routine, but is commented out. Maybe add this back in, or move to the PatchDynamicsMod?

https://github.com/NGEET/fates/blob/ac07ad4b9c2c07ae6ef0f5cf085b82848b4f33e9/fire/SFMainMod.F90#L908
```ruby
907      !shrink canopy to account for burnt section.     
908      !currentCohort%canopy_trim = min(currentCohort%canopy_trim,(1.0_r8-currentCohort%cfa)) 
```
If the canopy is trimmed, the next step/question becomes how to represent the branches consumed by this fire. Again, these trees are alive, but the removal of the branches has the capacity to create a delayed damage, and will take the trees off allometry. I need to look into the relationship between fire and branch consumption of live trees, but this seems to keep coming back to issue #413. Branch removal is an extreme form of altering the hydraulics of the tree, but also results in a flux to the atmosphere. This section of EdPatchDynamicsMod is where to account for that branch and sapwood flux. I am inclined, however, to leave this section as a flux of leaves at the moment and look more into the branch and sapwood consumption for living trees. Will chat with the Fire Sciences Research Lab in Missoula on this branch consumption point.

comments and other thoughts on this...?

As an aside, it would be useful to separate these fluxes by PFT as that information is useful for emissions inventories. May open an issue on this PFT flux separation.

**Comment by jkshuman** [comment]
For this specific issue, I removed the mimimim evaluation given that we already have a check on cfa within the fire mod. I also removed the flux for sapwood with the expectation that it will be added in when we add in branch consumption. Will open a separate issue specific to this task of branch consumption. I created a pull request with these changes. @rgknox @ckoven @rosiealice Let's check it over, and decide if we want to call this issue closed, and open a separate issue for the branch consumption.

To round out the previous details, we already include the proportion of the biomass allocated to each part for separation into the fuels for the fire routine. The parameter fates_CWD_frac can be updated in the parameter file to change these proportions. Currently they are: 0.045 (twigs), 0.075 (small branches), 0.21 (large branches), 0.67 (stem).

It would be worthwhile to have further discussion about how we handle branches for fire damage, hydraulics, respiration and movement of resources. Specifically, it may be worth discussing the importance of tapering, and if that will help with any of these issues.
